components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigration.kt

```diff
@@ -0,0 +1,216 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.support.migration
+
+import android.content.Context
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.async
+import mozilla.components.browser.storage.sync.PlacesBookmarksStorage
+import mozilla.components.browser.storage.sync.PlacesHistoryStorage
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.base.utils.SharedPreferencesCache
+import org.json.JSONArray
+import org.json.JSONObject
+import java.io.File
+import java.lang.Exception
+import java.lang.IllegalStateException
+import java.util.concurrent.Executors
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Supported Fennec migrations.
+ */
+enum class Migration {
+    HISTORY, BOOKMARKS, OPEN_TABS
+}
+
+/**
+ * Result of running a set of [Migration]s. Maps a [Migration] to a tuple of (migration version, [MigrationResult]).
+ */
+typealias MigrationsResult = Map<Migration, Pair<Int, MigrationResult>>
+
+// Describes ordering and versioning of migrations.
+private val migrations = listOf(
+    Migration.HISTORY to 1,
+    Migration.BOOKMARKS to 1,
+    Migration.OPEN_TABS to 1
+)
```

Comment:
I wonder if we can make this more transparent for `MigrationHistoryStore`. So that the store doesn't know about what migrations it actually runs. Our migration code would implement some `Migration` interface (or whatever)  and we register those with the migration store/manager thingy here. This allows us to more flexible "compose" a migration instead of hardcoding everything in the store. What do you think? I think it would be quite nice if Fenix (or another Fennec fork) could be in control what gets migrated this way (and we can land things in AC without immediately affecting Fenix with the next snapshot).

Reply:
I've updated `FennecMigration` constructor to take in a list of migrations and their versions, allowing Fenix to configure run whatever migrations it needs. I think that achieves what you describe, right?

This was also necessary to make things testable.